From b3f79d3b2602d3059d28f27c76e0f5af77666146 Mon Sep 17 00:00:00 2001
From: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
Date: Thu, 23 Jul 2026 09:54:56 +0800
Subject: [PATCH] mtk: hostapd: fix background primary channel backup issue

When the primary channel backup is out of date due to radar detection,
instead of exiting and leaving the background channel incorrect,
set it to the first primary channel.

Without this patch, the main radio channel switch will fail due to using
an incorrect background channel after background CAC is completed.

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
---
 src/ap/dfs.c | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

diff --git a/src/ap/dfs.c b/src/ap/dfs.c
index 65448c92b..960611abd 100644
--- a/src/ap/dfs.c
+++ b/src/ap/dfs.c
@@ -1357,10 +1357,11 @@ hostapd_dfs_background_expand(struct hostapd_iface *iface, int chan_width)
 	}
 
 	primary_backup = iface->radar_background.primary_backup;
-	if (chan->chan > primary_backup ||
+	if (!primary_backup || chan->chan > primary_backup ||
 	    primary_backup > chan->chan + expanded_width / 5 - 4) {
-		wpa_printf(MSG_ERROR, "DFS: Invalid backup primary ch: %d", primary_backup);
-		goto out;
+		wpa_printf(MSG_ERROR, "DFS: Invalid backup primary ch: %d change to %d",
+			   primary_backup, chan->chan);
+		primary_backup = chan->chan;
 	}
 
 	iface->radar_background.freq = 5000 +  5 * primary_backup;
@@ -1661,6 +1662,9 @@ hostapd_dfs_background_start_channel_switch(struct hostapd_iface *iface,
 		 * Just select a new random channel according to the
 		 * regulations for monitoring.
 		 */
+		/* the primary backup is out-of-date */
+		if (iface->radar_background.primary_backup)
+			iface->radar_background.primary_backup = 0;
 		hostapd_dfs_update_background_chain(iface);
 		return 0;
 	}
-- 
2.45.2

